### ISSUE TITLE
Requres reauthentication on SPA refresh

### ISSUE THREAD

**Issue by cschooley** [minbrowser/min#713]
Hi,

Awesome browser! Reporting this, but I intend to look at the issue further.

I downloaded and installed Min v1.9.0 on Linux Mint 18.1. It seems to work great for all SPA or non-trivial web interfaces I tested with (Kibana, various webmail, etc), but when using Min to browse to a local copy of ntopng (v.3.8.190206) I get prompted (by the SPA) to re-authenticate every 1 second.

Steps to reproduce:
* Log into the application (ntopng (v.3.8.190206)) using forms-based authentication (hosted at http://192.168.x.x:3000 * )
* The authentication protected pages are visible
* After about 1 second the login page is displayed without any user interaction

I am not able to reproduce this is Firefox, surf, or Chrome/Chromium. To me this sounds like the ntopng is refreshing it's display via XHR (it's got a UI full of graphs that update constantly) and that is causing the re-login cycle. * I mention I am hosting on an IP in case this breaks some kind of same origin stuff for XHR.

I tried to 'Allow all ads and trackers' from the Preferences (the used Developer | Reload Browser) in case ntopng was trying and failing to set some cookie, but that does not affect the behavior.

I just discovered the Inspect Browser feature and am familiar with debugging in Chrome/Chromium, so I'll will research and try to provide more detail on the refresh/authentication mechanism.

**Comment by cschooley** [comment]
The behavior observed (redirecting to login screen) is how this particular application handles errors (via callback), in this case running the app in Min I get an error in a function called footerRefresh that I do not get in other browsers tested. Will continue with the debugger to find the exact line of JavaScript that fails in Min.

**Comment by cschooley** [comment]
TL;DR: In this case Min is not sending cookies with the AJAX request if it's being called async from setInterval(). Not sure if this is by design or not.

Details: ntopng has a function called footerRefresh() that is called immediately when the page loads to populate the UI, then setInterval(footerRefresh, 3 * 1000) is called:

`
footerRefresh();  /* call immediately to give the UI a more responsive look */
setInterval(footerRefresh, 3 * 1000);  /* re-schedule every [interface-rate] seconds */
`

When I look in the Network tab in the Inspect Page (Chrome Debugger) in Min I see that the first (manual, non async) call to footerRefresh (network_load.lua?fid=0) correctly sends a cookie called 'session' that does what you'd think. The second time footerRefresh is called (again XHR to network_load.lua?fid=0) there are no cookies in the request.

Aside from the first invocation happening when the page loads (inline javascript) and the 2nd being called by setInterval() the call stacks are identical.

Please let me know if I can provide more information, hop on a screen share, try something else, etc. I'd like to help in any way possible.

**Comment by PalmerAL** [comment]
Thanks for reporting this, and for being so willing to help investigate it!

After installing ntopng, I can also see something similar. However, at least in my case, it doesn't always happen on the second call; instead, it will occur seemingly randomly at any point, from right after the page loads to after several minutes.

I also haven't been able to reproduce this in a standalone copy of Electron, using a `browserView` with the same options as Min. (I also tried disabling the `onBeforeRequest` listener in Min, and that had no effect). It's possible there's something else in Min that may be causing this, but I'm guessing it's a general electron issue and I just haven't been able to reproduce it there due to luck.

Unfortunately, I don't know very much about how cookies work in Electron, so I don't have a good idea of where to start investigating this. Given that this works in Chrome, the problem is probably in one of the [various patches](https://github.com/electron/electron/search?q=cookie&unscoped_q=cookie) that Electron adds on top of Chrome's cookie behavior, which leaves a lot of possibilities:

https://github.com/electron/electron/blob/434a6e3561fea806dd3f0b3888d8c2e20b650a6f/atom/browser/cookie_change_notifier.cc
https://github.com/electron/electron/blob/d53b51607c7b69bbc53c41b984963eef21d69f0b/atom/browser/api/atom_api_cookies.cc
https://github.com/electron/electron/blob/1918e76913ed550a3b2738ac541de48cb9db4afa/atom/browser/net/url_request_context_getter.cc

I'm not seeing anything that obviously looks like a problem in any of these, although as I mentioned, I don't really know enough about this to know for sure. It's probably worth adding some logging in a few of these places and see where the cookie is getting removed.

Alternatively, if we could come up with a reduced testcase for this, the Electron maintainers would probably have a better idea of how to solve it. Since you mentioned `setInterval`, I tried sending repeated requests, normally and then in an interval:
```
function makeRequest() {
    document.cookie = "session=test;";
    $.ajax({
        method: "get",
        url: window.location.toString()
    })
};
for (var i = 0; i < 5; i++) {
    makeRequest()
};
setInterval(makeRequest, 5000)
```
Doing that, the session cookie was persisted; in fact, the cookie set from `document.cookie` was duplicated with the cookie returned from the server using `set-cookie`, resulting in this header:
```
 Cookie: session=test; session=
```
Which is a [known issue](https://github.com/electron/electron/issues/16292) in Electron. (I suppose it's possible that that behavior is causing this issue somehow, although I don't see how that could be the case).

**Comment by cschooley** [comment]
Thanks for taking a look at this, tagging it, and the electron background. I'll take your suggestion and come up with a reduced test case for Min (to take ntopng out of the picture). Then I'll see if I can build a simple Electron test case and log/debug it assuming I can reproduce the issue. That should get me to the point where I can open a bug (or update the one linked) against Electron. I'll reference any upstream issue reports in this thread for tracking.
